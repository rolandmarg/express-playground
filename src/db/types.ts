export type Meeting = {
  id: number;
  title: string;
  startsAt: Date;
  endsAt: Date;
};

export type MeetingIn = Omit<Meeting, 'id'>;

export type User = {
  id: number;
  email: string;
  providers?: Provider[];
};

export type Provider = {
  id: number;
  providerName: string;
  providerId: string;
  email: string;
  accessToken: string;
  refreshToken?: string;
  photo?: string;
  displayName?: string;
  fullName?: string;
  user: User;
};

export type ProviderIn = Omit<Provider, 'id' | 'user'>;
