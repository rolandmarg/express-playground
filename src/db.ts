import { createConnection, Connection } from 'typeorm';
import env from './env';

import { User } from './entity/User';
import { Meeting } from './entity/Meeting';
import { Provider } from './entity/Provider';

export { User, Meeting, Provider };

let connection: Connection;

export const connect = async () => {
  connection = await createConnection({
    type: 'postgres',
    url: env.DB_URL,
    synchronize: env.NODE_ENV !== 'production',
    entities: [User, Meeting, Provider],
    logging: ['info'],
  });

  return connection;
};

export const close = async () => {
  return connection.close();
};
