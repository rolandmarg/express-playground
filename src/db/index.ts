import { createConnection, Connection } from 'typeorm';
import env from '../env';

import { User } from './entity/User';
import { Meeting } from './entity/Meeting';
import { Provider } from './entity/Provider';

export * from 'typeorm';
export * as meetingRepo from './repository/meeting';

export { User, Meeting, Provider };

let connection: Connection;
export const connect = async () => {
  connection = await createConnection({
    type: 'postgres',
    url: env.DB_URL,
    synchronize: env.NODE_ENV !== 'production',
    entities: [User, Meeting, Provider],
  });

  return connection;
};

export const close = async () => {
  return connection.close();
};
