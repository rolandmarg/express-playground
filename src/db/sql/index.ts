import { QueryFile } from 'pg-promise';

import { join as joinPath } from 'path';

function sql(file: string): QueryFile {
  const fullPath: string = joinPath(__dirname, file);

  const qf = new QueryFile(fullPath, { minify: true });

  if (qf.error) {
    console.error(qf.error);
  }

  return qf;
}

export const users = {
  create: sql('users/create.sql'),
  drop: sql('users/drop.sql'),
  delete: sql('users/delete.sql'),
  insert: sql('users/insert.sql'),
  insertOrIgnore: sql('users/insertOrIgnore.sql'),
  selectByEmail: sql('users/selectByEmail.sql'),
};

export const providers = {
  create: sql('providers/create.sql'),
  drop: sql('providers/drop.sql'),
  delete: sql('providers/delete.sql'),
  selectByEmail: sql('providers/selectByEmail.sql'),
};

export const meetings = {
  create: sql('meetings/create.sql'),
  drop: sql('meetings/drop.sql'),
  truncate: sql('meetings/truncate.sql'),
  insert: sql('meetings/insert.sql'),
  select: sql('meetings/select.sql'),
  searchByCursor: sql('meetings/searchByCursor.sql'),
  searchByDate: sql('meetings/searchByDate.sql'),
  selectFirst: sql('meetings/selectFirst.sql'),
  selectLast: sql('meetings/selectLast.sql'),
};

export default {
  users,
  providers,
  meetings,
};
