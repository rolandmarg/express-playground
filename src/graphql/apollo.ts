import { Request, Response } from 'express';
import { ApolloServer } from 'apollo-server-express';
import { typeDefs } from './typeDefs';
import { resolvers } from './resolvers';
import { AuthDirective } from './authDirective';
import { User } from '../entity';
import { AppError } from '../utils';

export interface IContext {
  req: Request;
  res: Response;
  user?: User;
}

const apolloServer = new ApolloServer({
  typeDefs,
  resolvers,
  schemaDirectives: { auth: AuthDirective },
  formatError: (err) => {
    if (err.originalError instanceof AppError) {
      return err;
    } else {
      console.error(err);
      return new AppError();
    }
  },
});

export const apollo = () => apolloServer.getMiddleware();
