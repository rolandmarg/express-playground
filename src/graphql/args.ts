import { ArgsType, Int, Field } from 'type-graphql';
import { Min } from 'class-validator';

@ArgsType()
export class ConnectionArgs {
  @Field(() => Int, { defaultValue: 10 })
  @Min(1)
  first!: number;

  @Field({ nullable: true })
  after?: string;
}
