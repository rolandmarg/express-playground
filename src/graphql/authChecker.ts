import { AuthChecker } from 'type-graphql';
import { Context } from './context';
import { authRequest } from '../auth/helpers';

export const customAuthChecker: AuthChecker<Context> = async ({ context }) => {
  const user = await authRequest(context.req);

  context.currentUser = user;

  return !!context.currentUser;
};
