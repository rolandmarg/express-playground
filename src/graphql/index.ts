export * from './apollo';

export * as queries from './query';
export * as mutations from './mutation';
export * as fragments from './fragments';
