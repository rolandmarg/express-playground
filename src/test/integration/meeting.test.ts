import { createTestClient } from 'apollo-server-testing';
import { Repository, getRepository } from 'typeorm';
import { connect, close, Meeting } from '../../db';
import { apolloServer, queries, mutations } from '../../graphql';

jest.mock('../../env', () => ({
  DB_URL: 'postgresql://rem@localhost:5432/midnightest',
}));

const { query, mutate } = createTestClient(apolloServer);

describe('Meeting operations', () => {
  let meetingRepo: Repository<Meeting>;

  beforeAll(async () => {
    await connect();
    meetingRepo = getRepository(Meeting);
  });

  afterAll(async () => {
    await close();
  });

  beforeEach(async () => {
    await meetingRepo.clear();
  });

  it('should fetch freshly created meetings from database', async () => {
    const title = 'testMeeting';
    await meetingRepo.save({ title, startsAt: new Date(), endsAt: new Date() });

    const res = await query({ query: queries.meetings });

    expect(res).toMatchSnapshot({
      data: {
        meetings: [
          {
            id: expect.any(String),
            endsAt: expect.any(String),
            startsAt: expect.any(String),
            title,
          },
        ],
      },
    });
  });

  it('should fetch empty list of meetings', async () => {
    const res = await query({ query: queries.meetings });

    expect(res).toMatchSnapshot();
  });

  it('should fetch specific meeting', async () => {
    const title = 'testMeeting';
    const meeting = await meetingRepo.save({
      title,
      startsAt: new Date(),
      endsAt: new Date(),
    });

    const res = await query({
      query: queries.meeting,
      variables: { id: meeting.id },
    });

    expect(res).toMatchSnapshot({
      data: {
        meeting: {
          id: expect.any(String),
          endsAt: expect.any(String),
          startsAt: expect.any(String),
          title,
        },
      },
    });
  });

  it('should create a new meeting', async () => {
    const title = 'testMeeting';
    const res = await mutate({
      mutation: mutations.createMeeting,
      variables: {
        title,
        startsAt: new Date().toISOString(),
        endsAt: new Date().toISOString(),
      },
    });

    expect(res).toMatchSnapshot({
      data: {
        createMeeting: {
          meeting: {
            id: expect.any(String),
            endsAt: expect.any(String),
            startsAt: expect.any(String),
            title,
          },
        },
      },
    });

    const newMeeting = await meetingRepo.findOne({ title });

    expect(newMeeting).toBeDefined();
  });
});
