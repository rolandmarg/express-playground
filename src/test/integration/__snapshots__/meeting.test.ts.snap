// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meeting operations prev page should be true after fetching from specific date 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [
        Object {
          "cursor": Any<String>,
          "node": Object {
            "endsAt": Any<String>,
            "id": Any<String>,
            "startsAt": Any<String>,
            "title": "testMeeting",
          },
        },
        Object {
          "cursor": Any<String>,
          "node": Object {
            "endsAt": Any<String>,
            "id": Any<String>,
            "startsAt": Any<String>,
            "title": "testMeeting",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations prev page should be true when getting second meeting 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [
        Object {
          "cursor": Any<String>,
          "node": Object {
            "endsAt": Any<String>,
            "id": Any<String>,
            "startsAt": Any<String>,
            "title": "testMeeting",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations prev/next pages should be true when getting second from many meetings 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [
        Object {
          "cursor": Any<String>,
          "node": Object {
            "endsAt": Any<String>,
            "id": Any<String>,
            "startsAt": Any<String>,
            "title": "testMeeting",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations should create a new meeting 1`] = `
Object {
  "data": Object {
    "createMeeting": Object {
      "endsAt": Any<String>,
      "id": Any<String>,
      "startsAt": Any<String>,
      "title": "testMeeting",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations should delete a meeting 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [],
      "pageInfo": Object {
        "endCursor": null,
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations should fetch empty list of meetings 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [],
      "pageInfo": Object {
        "endCursor": null,
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations should fetch freshly created meeting from database 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [
        Object {
          "cursor": Any<String>,
          "node": Object {
            "endsAt": Any<String>,
            "id": Any<String>,
            "startsAt": Any<String>,
            "title": "testMeeting",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations should have different end/start cursors & next page true on two meetings 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [
        Object {
          "cursor": Any<String>,
          "node": Object {
            "endsAt": Any<String>,
            "id": Any<String>,
            "startsAt": Any<String>,
            "title": "testMeeting",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Meeting operations should have same end/start cursors & next/previous pages false on one meeting 1`] = `
Object {
  "data": Object {
    "meetings": Object {
      "edges": Array [
        Object {
          "cursor": Any<String>,
          "node": Object {
            "endsAt": Any<String>,
            "id": Any<String>,
            "startsAt": Any<String>,
            "title": "testMeeting",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
